File: backend/internal/model/schema.go
package model

// ColumnInfo represents metadata about a database column
type ColumnInfo struct {
	ColumnName             string  `json:"column_name"`
	DataType               string  `json:"data_type"`
	IsNullable             string  `json:"is_nullable"`
	ColumnDefault          *string `json:"column_default"`
	CharacterMaximumLength *int    `json:"character_maximum_length"`
	NumericPrecision       *int    `json:"numeric_precision"`
	UdtName                string  `json:"udt_name"`
	IsPrimaryKey           bool    `json:"is_primary_key"`
	IsForeignKey           bool    `json:"is_foreign_key"`
}

// ForeignKeyInfo represents a foreign key relationship
type ForeignKeyInfo struct {
	TableName         string `json:"table_name"`
	ColumnName        string `json:"column_name"`
	ForeignTableName  string `json:"foreign_table_name"`
	ForeignColumnName string `json:"foreign_column_name"`
	ConstraintName    string `json:"constraint_name"`
}

// TableSchema represents the complete schema for a single table
type TableSchema struct {
	TableName     string           `json:"table_name"`
	Columns       []ColumnInfo     `json:"columns"`
	ForeignKeys   []ForeignKeyInfo `json:"foreign_keys"`
	PrimaryKey    string           `json:"primary_key"`
	PrimaryKeys   []string         `json:"primary_keys"`
	HasTimestamps bool             `json:"has_timestamps"`
}

// DatabaseSchema represents the complete discovered database schema
type DatabaseSchema struct {
	Tables       []TableSchema `json:"tables"`
	DiscoveredAt string        `json:"discovered_at"`
	Version      int           `json:"version"`
}

// ColumnType represents the UI type for rendering form fields
type ColumnType string

const (
	TypeText        ColumnType = "text"
	TypeNumber      ColumnType = "number"
	TypeBoolean     ColumnType = "boolean"
	TypeDate        ColumnType = "date"
	TypeDateTime    ColumnType = "datetime"
	TypeJSON        ColumnType = "json"
	TypeUUID        ColumnType = "uuid"
	TypeEmail       ColumnType = "email"
	TypeURL         ColumnType = "url"
	TypeTextArea    ColumnType = "textarea"
	TypeImage       ColumnType = "image"
	TypeSelect      ColumnType = "select"
	TypeMultiSelect ColumnType = "multiselect"
	TypeCode        ColumnType = "code"
)

File: frontend/src/types/schema.ts
// Core schema types discovered at runtime

export interface ColumnInfo {
  column_name: string;
  data_type: string;
  is_nullable: string;
  column_default: string | null;
  character_maximum_length: number | null;
  numeric_precision: number | null;
  udt_name: string;
  is_primary_key: boolean;
  is_foreign_key: boolean;
}

export interface ForeignKeyInfo {
  table_name: string;
  column_name: string;
  foreign_table_name: string;
  foreign_column_name: string;
  constraint_name: string;
}

export interface TableSchema {
  table_name: string;
  columns: ColumnInfo[];
  foreign_keys: ForeignKeyInfo[];
  primary_key: string;
  primary_keys?: string[];
  has_timestamps: boolean;
}

export interface DatabaseSchema {
  tables: TableSchema[];
  discovered_at: string;
  version: number;
}

// UI Field types mapped from PostgreSQL types
export type FieldType = 
  | 'text'
  | 'number'
  | 'boolean'
  | 'date'
  | 'datetime'
  | 'json'
  | 'uuid'
  | 'email'
  | 'url'
  | 'textarea'
  | 'image'
  | 'select'
  | 'multiselect'
  | 'code';

export interface FieldConfig {
  name: string;
  label: string;
  type: FieldType;
  required: boolean;
  nullable: boolean;
  defaultValue?: any;
  maxLength?: number;
  precision?: number;
  foreignKey?: ForeignKeyInfo;
  options?: { label: string; value: any }[];
  placeholder?: string;
  description?: string;
}

// API Response types
export interface ApiResponse<T = any> {
  success: boolean;
  data?: T;
  error?: string;
  details?: string;
  message?: string;
  meta?: PaginatedMeta;
}

export interface PaginatedMeta {
  data: Record<string, any>[];
  total: number;
  page: number;
  page_size: number;
  total_pages: number;
}

// Auth types
export interface User {
  id: string;
  email: string;
  role: 'admin' | 'student' | 'editor' | 'viewer' | 'authenticated';
  name?: string;
  avatar_url?: string;
  student_id?: string;
}

export interface AuthTokens {
  access_token: string;
  refresh_token: string;
  user: User;
}

export interface StudentChangeRequest {
  id: string;
  requester_user_id: string;
  requester_name?: string;
  requester_roll_no?: string;
  student_id?: string;
  request_type: 'claim_student' | 'create' | 'update' | 'delete' | 'link' | 'unlink';
  target_table: string;
  target_key?: Record<string, any>;
  current_data?: Record<string, any>;
  proposed_data: Record<string, any>;
  status: 'pending' | 'approved' | 'rejected';
  admin_note?: string;
  reviewed_by?: string;
  reviewed_at?: string;
  created_at: string;
  updated_at: string;
}

export interface StudentWorkspace {
  profile: User;
  student?: Record<string, any>;
  linked_records: Record<string, Record<string, any>[]>;
  references: Record<string, Record<string, any>[]>;
  requests: StudentChangeRequest[];
}

export interface StudentSearchResult {
  id: string;
  name: string;
  roll_no: string;
  image_url?: string | null;
  batch_id?: string | null;
}

export interface StudentContributor {
  id: string;
  name: string;
  roll_no: string;
  image_url?: string | null;
  role?: string | null;
}

export interface AccessUser {
  user_id?: string;
  student_id: string;
  roll_no: string;
  name: string;
  login_email: string;
  role: 'admin' | 'student';
  has_account: boolean;
}

export interface ProvisionUserResult {
  student_id: string;
  roll_no: string;
  login_email: string;
  role: 'admin' | 'student';
  status: 'created' | 'updated' | 'failed';
  message?: string;
  user_id?: string;
}

export interface ProvisionResult {
  created: number;
  updated: number;
  failed: number;
  users: ProvisionUserResult[];
}

// Table view config
export interface TableFilter {
  column: string;
  operator: 'eq' | 'ne' | 'gt' | 'lt' | 'like';
  value: string;
}

export interface TableSort {
  column: string;
  direction: 'ASC' | 'DESC';
}

export interface TableViewConfig {
  pageSize: number;
  sort: TableSort;
  filters: TableFilter[];
  visibleColumns: string[];
}
